src/poc/miner_poc_grpc_client_statem.erl

```diff
@@ -0,0 +1,1080 @@
+) ->
+    Req = #gateway_poc_check_challenge_target_req_v1_pb{
+        address = SelfPubKeyBin,
+        challenger = ChallengerPubKeyBin,
+        block_hash = BlockHash,
+        onion_key_hash = OnionKeyHash,
+        height = ChallengeHeight,
+        notifier = ChallengerPubKeyBin,
+        notifier_sig = ChallengerSig,
+        challengee_sig = <<>>
+    },
+    ReqEncoded = gateway_miner_client_pb:encode_msg(
+        Req, gateway_poc_check_challenge_target_req_v1_pb
+    ),
+    Req#gateway_poc_check_challenge_target_req_v1_pb{challengee_sig = SelfSigFun(ReqEncoded)}.
+
+-spec process_unary_response(grpc_client_custom:unary_response()) -> unary_result().
+process_unary_response(
+    {ok, #{
+        http_status := 200,
+        result := #gateway_resp_v1_pb{
+            msg = {success_resp, _Payload}, height = Height, signature = Sig
+        }
+    }}
+) ->
+    {ok, #{height => Height, signature => Sig}};
+process_unary_response(
+    {ok, #{
+        http_status := 200,
+        result := #gateway_resp_v1_pb{msg = {error_resp, Details}, height = Height, signature = Sig}
+    }}
+) ->
+    #gateway_error_resp_pb{error = ErrorReason} = Details,
+    {error, ErrorReason, #{height => Height, signature => Sig}};
+process_unary_response(
+    {ok, #{
+        http_status := 200,
+        result := #gateway_resp_v1_pb{msg = {_RespType, Payload}, height = Height, signature = Sig}
+    }}
+) ->
+    {ok, Payload, #{height => Height, signature => Sig}};
+process_unary_response({error, _}) ->
+    lager:warning("grpc client error response", []),
+    {error, grpc_request_failed}.
+
+-spec get_uri_for_challenger(
+    OnionKeyHash::binary(),
+    Connection::grpc_client_custom:connection()) ->
+    {ok, {string(), pos_integer()}} | {error, any()}.
+get_uri_for_challenger(OnionKeyHash, Connection) ->
+    Req = build_poc_challenger_req(OnionKeyHash),
+    case send_grpc_unary_req(Connection, Req, 'poc_key_to_public_uri') of
+        {ok, #gateway_public_routing_data_resp_v1_pb{public_uri = URIData}, _Req2Details} ->
+            #routing_address_pb{uri = URI, pub_key = _PubKey} = URIData,
+            #{host := IP, port := Port} = uri_string:parse(binary_to_list(URI)),
+            {ok, {IP, Port}};
+        {error, Reason, _Details} ->
+            {error, Reason};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec process_check_target_reqs() -> ok.
+process_check_target_reqs() ->
+    {ok, POCTimeOut} = application:get_env(miner, poc_timeout),
+    %% convert poc timeout in blocks to equiv in seconds
+    %% use to determine when the POC would end
+    %% if the cached request has remained in the cache
+    %% for 80% of the timeout period, then give up on it & remove
+    %% we use 80% as we need to give the POC time to run,
+    %% no point getting the onion and broadcasting if no time
+    %% left for it to complete
+    POCTimeOutSecs = (POCTimeOut * 0.8 * 60),
+    CurTSInSecs = erlang:system_time(second),
```

Comment:
This is a bit risky as it assumes an accurate system clock, it might be better to trust the time from the validator?

Reply:
the calculation is relative tho, the times being compared are both system time